User option for deleting/disabing forms and deleting submissions

rohitdatta opened the pull request:
**Changes proposed in this pull request:**

* Allow form owners to delete individual forms
* Allow form owners to delete individual submissions
* Allow form owners to disable forms

**Have you made sure to add:**
- [X] Tests
- [ ] Documentation

**Screenshots and GIFs**
<img width="1271" alt="screen shot 2016-03-15 at 1 20 45 am" src="https://cloud.githubusercontent.com/assets/6260066/13768429/62013026-ea4c-11e5-9772-88b070b0038d.png">

**Deploy notes**

~~No migration/upgrade necessary~~
A new column was added for disabled forms, so now must run `heroku run python manage.py db upgrade --app formspree-asm`

**Any Additional Information**
This PR has changed significantly since proposed. Please look below to see updates.

fiatjaf commented on formspree/forms/models.py:
```diff
@@ -312,3 +312,10 @@ def __init__(self, form_id):
+        try:
+            return Submission.query.filter_by(id=submissionid).first()
+        except IndexError:
+            return None
```
is an `IndexError` ever raised here? `.first()` normally returns `None` when no records are found in the query. It is not the same as calling `.all()[0]`.

fiatjaf commented on formspree/routes.py:
```diff
@@ -30,6 +30,8 @@ def configure_routes(app):
     app.add_url_rule('/forms/<hashid>/', 'form-submissions', view_func=forms.views.form_submissions, methods=['GET'])
     app.add_url_rule('/forms/<hashid>.<format>', 'form-submissions', view_func=forms.views.form_submissions, methods=['GET'])
+    app.add_url_rule('/forms/<hashid>/delete', 'form-deletion', view_func=forms.views.form_deletion, methods=['GET'])
+    app.add_url_rule('/forms/<hashid>/<submissionid>/delete', 'submission-deletion', view_func=forms.views.submission_deletion, methods=['GET'])
```
These endpoints shouldn't be available through `GET` requests. That's an antipattern and opens easy vulnerabilities such as someone that knows your form hashid making you click on a link that then deletes it.

rohitdatta replied:
Even with the authentication methods below, should we have a confirmation screen just for the users sake? I think we should for form deletion, but if we do for submission deletion do you think it becomes too cumbersome?

fiatjaf replied:
For submission deletion I think it becomes too cumbersome and I guess unnecessary, but I like it for form deletion.

colevscode replied:
I don't think we should actually delete the form. It should be marked as deleted and then displayed in a (possibly collapsed) section below the active forms. Because the form is never really deleted, it shouldn't be necessary to provide a confirmation.

fiatjaf commented on formspree/forms/views.py:
```diff
@@ -366,3 +366,48 @@ def form_submissions(hashid, format=None):
+
+@login_required
+def submission_deletion(hashid, submissionid):
+    submission = Submission.get_by_submissionid(submissionid)
```
The `get_by_submissionid` method is kinda useless boilerplate (I don't know if these words are offensive, but I didn't meant them to be).

The exact same behavior here can be achieved with

```python
submission = Submission.query.get(submissionid)
```

fiatjaf commented on formspree/templates/forms/list.html:
```diff
@@ -35,6 +35,9 @@
       </td>
+		<td>
+		<a href="{{ url_for('form-deletion', hashid=form.hashid) }}">Delete</a>
+		</td>
```
Since we can't use `GET` endpoints, I think the simplest solution is to use a `POST` endpoint and replace this `<a>` with an inline form with just a button -- I think this is also vulnerable, but a little less (and I think it is an approach being used in other parts of Formspree).

rohitdatta replied:
Yep I wasn't sure about this. I added these methods and said they needed to be logged in to perform them and then if they are logged in they need to actually be the form owner in order to delete. But I wasn't really sure between a GET and POST request since I guess you could say we're posting data but I was POSTing through the URL unless you think I should send that data through a form and have the hashid and submission id be the data that was posted.

Thank you very much for your other points also! I wasn't really sure about that stuff, so good to know the correct syntax.

colevscode commented:
Having a delete button on the same line is fine but we should use a small
trash can icon. I think we must be using fontawesome or another font icon
set that contains a trash can icon.
On Mar 15, 2016 1:23 AM, "Rohit Datta" <notifications@github.com> wrote:

> *Changes proposed in this pull request:*
>
>    - Allow form owners to delete individual forms
>    - Allow form owners to delete individual submissions
>
> *Have you made sure to add:*
>
>    - Tests
>    - Documentation
>
> *Screenshots and GIFs*
> [image: screen shot 2016-03-15 at 1 20 45 am]
> <https://cloud.githubusercontent.com/assets/6260066/13768429/62013026-ea4c-11e5-9772-88b070b0038d.png>
>
> *Deploy notes*
>
> No migration/upgrade necessary
>
> *Any Additional Information*
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/formspree/formspree/pull/84
> Commit Summary
>
>    - Add option for deleting form and its submissions
>    - Allow for deletion of submissions
>    - Make sure that only authorized users can delete
>    - Add test suite for delete forms and submissions
>
> File Changes
>
>    - *M* formspree/forms/models.py
>    <https://github.com/formspree/formspree/pull/84/files#diff-0> (7)
>    - *M* formspree/forms/views.py
>    <https://github.com/formspree/formspree/pull/84/files#diff-1> (45)
>    - *M* formspree/routes.py
>    <https://github.com/formspree/formspree/pull/84/files#diff-2> (2)
>    - *M* formspree/templates/forms/list.html
>    <https://github.com/formspree/formspree/pull/84/files#diff-3> (3)
>    - *M* formspree/templates/forms/submissions.html
>    <https://github.com/formspree/formspree/pull/84/files#diff-4> (1)
>    - *M* tests/test_users.py
>    <https://github.com/formspree/formspree/pull/84/files#diff-5> (95)
>
> Patch Links:
>
>    - https://github.com/formspree/formspree/pull/84.patch
>    - https://github.com/formspree/formspree/pull/84.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/formspree/formspree/pull/84
>

rohitdatta commented:
Yeah an icon makes more sense. What do you think about a confirmation screen? We wouldn't want users to accidentally delete but at the same time, it shouldn't be a cumbersome user experience to delete

colevscode commented:
As I mentioned above, I don't think the forms should actually be deleted. They should just be marked "deleted" and then moved to a deleted section below the active form entries. "deleted" forms should then also stop sending emails.

Part of the reason I think this is the right course of action is, consider forms that were created before the user created an account. These will use an email address instead of a pseudorandom string for the form's action. Deleting a form like this would be the same as unconfirming it. Following submissions would generate a new confirmation notice. I think that's unexpected.

What should we show to the visitor that tries to submit a form that has been deactivated (deleted)? Perhaps they should see an error message that says the form has been deleted or disabled.

colevscode commented:
By marked "deleted" I mean we should add a deleted field to the database, and then the API endpoint will just set this field to true. This will require a migration.

rohitdatta commented:
Well oops, I didn't notice any of these comments before I pushed the changes I made tonight. I'll redo this to account for a "disabled" state on the form. One thing I might suggest however would be to still allow for a user to delete after they have disabled. Because the main reason why I wrote this was because some Gold customers had written in complaining that they were seeing test forms that they created.

I think the disabled forms would be good for people to see forms that they didn't want to receive any submissions and we could allow them to reenable it, but they could also delete it because it was just a test form they created. For submissions, would just deletion make sense? Because I think the main reason for deleting a submission is for getting rid of the test submissions you made while testing out a form.

rohitdatta commented:
Now that I think about it for a bit, wouldn't it make more sense to have both? A disabled state where it's hidden from view (and have the behavior you described), but that could be re-enabled, but also a complete delete button, where the user could just get rid of something if they never want it again. It wouldn't even add that much to the user dashboard, since we could just have a lock for disabled/enabled status and a trash can for delete, so it wouldn't take up that much space, but it would give more options.

rohitdatta commented:
I added the option for toggling forms on and off. However if you want to delete it permanently so you never have to see it again, that's possible, and it will show a modal asking you to confirm deletion.

If you want to only disable the form, that is also possible. A disabled form will return an error page to a user that attempts to submit forms because it is not accepting submissions. The form owner can re-enable it from their dashboard if/when they want to accept submissions again.

I added a column in the table for form called disabled, so to use this feature now, it will require a DB migration. It is live on [my Heroku](formspree-rohit.herokuapp.com) if you would like to see it in action. I can manually upgrade someone to Gold if they would like to test.

fiatjaf commented:
Please upgrade me.

rohitdatta commented:
Done.

fiatjaf commented:
That's a beautiful piece of software.

fiatjaf commented on tests/test_users.py:
```diff
@@ -189,6 +189,188 @@ def test_form_creation(self):
+        self.assertEqual(0, Submission.query.count())
+
+        # increase the submission limit
+        settings.ARCHIVED_SUBMISSIONS_LIMIT = 10
```
I know this `ARCHIVED_SUBMISSIONS_LIMIT` is a mess and I am the responsible for it, but shouldn't it be set to the previous value after this test?

rohitdatta replied:
Oh yes this is true. When I originally looked at the setup code I guess I thought it would tear everything down again and set it up, but now I realize that it's only setting up the database each time, not the submission limits and such, so I should change that because that's going to be very hard to debug otherwise.

fiatjaf commented:
I tried to submit the POST request for deleting the from an external website (by copying the form at http://formspree-rohit.herokuapp.com/dashboard#delete-ajpxzymy), but it failed with error 500. I don't know what happened, but I think it is fine.

Anyway, if this is a vulnerability there are probably other vulnerabilities and we can solve them later.

rohitdatta commented:
That actually is a problem I should fix. Currently it is checking for two ways to ensure that you can't delete a form you don't have access to. First you need to be logged in otherwise you get redirected to login. If you're logged in, it checks the current user to the form owner and rejects it if you're not the owner.

Let me try to find this and fix it. How exactly did you attempt to delete the form? Was it just a POST request sent from a random site that had the required form data for deletion. If so, we could check the referrer header (although a malicious attacker could spoof this). However technically if the user is logged in, it's still OK for them to delete their form even if they're trying to do it through a non-standard way

rohitdatta commented:
But if I do it locally I can see the stack trace so I would like to try that and see if we can display an error page instead

rohitdatta commented:
So I'm sending POST requests using [Postman](https://www.getpostman.com/) using bogus referer headers and sending a POST to both my localhost and the Heroku app and both times it redirects me to the login page.

fiatjaf commented:
An attacker cannot spoof this, because the attack would require the user to be logged in, so the attacker must fool a Formspree user to submit himself a form to Formspree from his site, let's say. I guess verifying the referrer solves all the problems.

rohitdatta commented:
Oh I see what you're saying where we could technically be subject to XSS
and CSRF. Checking the referrer should solve that since I don't think those
can be spoofed with those attacks.
On Thu, Mar 17, 2016 at 11:18 AM Giovanni T. Parra <notifications@github.com>
wrote:

> An attacker cannot spoof this, because the attack would require the user
> to be logged in, so the attacker must fool a Formspree user to submit
> himself a form to Formspree from his site, let's say. I guess verifying the
> referrer solves all the problems.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/formspree/formspree/pull/84#issuecomment-197928313>
>

rohitdatta commented on formspree/forms/views.py:
```diff
@@ -387,6 +387,14 @@ def form_toggle():
+    # check that this request came from user dashboard to prevent XSS and CSRF
+    referrer = referrer_to_baseurl(flask.request.referrer)
+    service = referrer_to_baseurl(settings.SERVICE_URL)
+    if referrer != service:
```
I'm just saying here that see if the referrer is from the formspree.io domain. It's messy to check the exact URL because there may be arguments attached to the end sometimes and `/dashboard` and `/forms` are actually the same thing, but any request from formspree.io (or the SERVICE_URL) should be OK.

fiatjaf replied:
That's good. It is unnecessary to check for full URLs.

fiatjaf commented:
Ok, let's merge this.

rohitdatta commented:
Alright I'm going through now fixing the merge conflicts since this was started before a ton of the changes so it's got a lot of stuff I need to fix.

rohitdatta commented:
Everything should be fixed now @fiatjaf, but I'm not really sure how to do DB migrations without taking down the server, so I'd prefer for you to do the deployment so I don't bring down Formspree.

rohitdatta commented:
Actually, don't merge this as yet. I may have messed up the DB migration stuff that I think will cause issues when we attempt to deploy.

rohitdatta commented:
OK, I think I fixed it, but I'm not completely sure. Would you mind taking a look and checking over the migrations and making sure it's ok.

rohitdatta commented:
I confirmed that this works by setting up a new Heroku instance that mirrored the Formspree production application, and then upgrading to this commit and everything worked fine, so I'm going to merge and deploy this.

colevscode commented:
@rohitdatta Sorry I dropped out of this thread. Great work shipping these changes. Generally I like them though I still think we don't need both the trash can and the lock. Personally I think if you trash a form it should just go into the deleted section, which should be hidden by default. If users complain I think we could consider permanently deleting things from the trash. 

I'm making a fuss about this just because I think the two icons create a less clean look. But it's not such a big deal that we should prioritize this. I can try to make a PR later with my design ideas.
